_posts/2020-02-17-constructor-injection.md

```diff
@@ -0,0 +1,189 @@
+```java
+@Component
+public class Cake {
+
+	private static Logger LOGGER = LoggerFactory.getLogger(Cake.class);
+
+	private Flavor flavor;
+
+	public Cake(Flavor flavor) throws IllegalAccessException {
+		
+		//check if the required dependency is not null
+		if (flavor != null) {
+			this.flavor = flavor;
+			LOGGER.info("Flavor from Constructor Injection : " + flavor);
+		} else {
+			throw new IllegalArgumentException("Cake cannot be created with null flavor object");
+		}
+	}
+	
+	...
+}
+```
+
+In setter injection, these dependencies are initialized using setter methods only after the constructor invocation. Using setter injection, we can override the existing values of the object and provide different values. This kind of injection is preferable when we have optional dependencies.
+
+Constructor injection is extremely useful since we do not have to write separate business logic everywhere to check if the all the required dependencies are loaded thus simplifying code complexity.
+
+## Identifying Code smells
+
+Constructor injection can help us identify if our bean is dependent many other classes. If our constructor has large number of arguments, then we can be sure that our class has too many responsibilities. It is not a sign of quality code implying it should be refactored to better address proper separation of concerns.
+
+Constructor injection can help us find out circular dependencies in the code. If it exists, Spring would throw `BeanCurrentlyInCreationException` . This can help the developer identify classes that have circular dependency.
+
+## Preventing NullPointerException
+
+Constructor injection simplifies writing unit test cases for any spring application. Using `Mockito`, we can create complete mock objects with required dependencies.
+
+You may ask why can't I mock objects created with setter injection. **Mockito unlike Spring fails with a null pointer exception if it cannot inject any of the dependencies**.
+
+Let us consider an example test case for Cake class that uses setter injection. Here the test works fine. 
+```java
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringBootTest(classes = ExampleApplicationSI.class)
+public class TestCakeClassSetterInjection {
+
+	@Autowired
+	Cake cake;
+
+	@Test
+	public void testSetterInjection() {
+		String testColor = cake.getFlavor().getColor();
+		Assert.assertEquals(testColor, " White ");
+		String toppingsName=cake.getToppings().getToppingName();
+		
+		//check if the dependency is not null
+		if(toppingsName!=null) {
+			Assert.assertEquals(toppingsName, "gems");
+		}
+		
+		...
+}
+``` 
+In future if someone adds additional dependencies to this class, the unit test silently fails with a null pointer exception. It is very difficult to find the reason because the new dependencies is not within the visibility for Cake’s client classes.
```

Comment:
I don't understand what you're saying with this paragraph and the code example above. Where do I add another dependency? Why is not in `Cake` visibility? 

I suggest you create a code example that really fails, explain why, and then show a code example that works fine. My problem with understanding is probably that the test case above "works fine" just as you said.

Reply:
I think now the newly added example will explain. Please let me know if that is okay.